### PR TITLE
Development 1387

### PR DESCRIPTION
**Pull request by monshri**
# Description

Please include a summary of the change, motivation and which issue is fixed. Any dependencies changes should also be included.

Implemented Poisoning Attack - Poison Frog in PyTorch Framework.

The implementation includes the below major changes:
- Creating Evaluation hooks for getting activation from intermediate layers implemented in _make_model_wrapper() in art/estimators/classification/pytorch.py. The motivation behind this change is that the current implementation of `get_layers` in _make_model_wrapper() the function can only infer the internal layers if the input model is of type `nn.Sequential`, otherwise, it will only return the logit layer. For Poison Frog implementation, it was needed that we get activations from intermediate layers and they stay connected to the computation graph for backward pass.
- A new custom_loss_gradient function especially implemented for calculating the gradients for Poison Frogs.
- A new jupyter notebook for running Poison Frogs on CIFAR10 dataset.

Fixes # (issue)
https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/1387

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
Could you please remove the *.pt files from the pull request and replace the models in the notebooks with downloads from https://www.dropbox.com/s/vmfgnkhjub47fy0/ciafr10_alexnet_pytorch.pt?dl=0 and/or https://www.dropbox.com/s/re6hd3w4m3rmfvh/ciafr10_alexnet_pytorch_linear.pt?dl=0

File: art/attacks/poisoning/feature_collision_attack.py [hunk 1]
```diff
@@ -31,6 +31,9 @@
 from art.estimators import BaseEstimator, NeuralNetworkMixin
 from art.estimators.classification.classifier import ClassifierMixin
 from art.estimators.classification.keras import KerasClassifier
+from art.estimators.classification.pytorch import PyTorchClassifier
+import matplotlib.pyplot as plt
```
Comment (beat-buesser) [inline art/attacks/poisoning/feature_collision_attack.py:35]:
```suggestion
```

File: art/attacks/poisoning/feature_collision_attack.py [hunk 2]
```diff
@@ -63,7 +66,7 @@ class FeatureCollisionAttack(PoisoningAttackWhiteBox):
         "verbose",
     ]
 
-    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin, ClassifierMixin, KerasClassifier)
+    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin, ClassifierMixin, PyTorchClassifier)
```
Comment (beat-buesser) [inline art/attacks/poisoning/feature_collision_attack.py:69]:
```suggestion
    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin, ClassifierMixin)
```

File: art/attacks/poisoning/feature_collision_attack.py [hunk 3]
```diff
@@ -110,12 +113,20 @@ def __init__(
...
+            self.poison_placeholder, self.poison_feature_rep = self.estimator.get_activations(
+                self.target, self.feature_layer, 1, framework=True
+            )
+        elif isinstance(self.estimator, PyTorchClassifier):
+            self.target_placeholder, self.target_feature_rep = self.estimator.get_activations(
+                self.target, self.feature_layer, 1, framework=True
+            )
+            self.poison_placeholder, self.poison_feature_rep = self.estimator.get_activations(
+                self.target, self.feature_layer, 1, framework=True
+            )
```
Comment (beat-buesser) [inline art/attacks/poisoning/feature_collision_attack.py:129]:
```suggestion
            )
        else:
            raise ValueError("Type of estimator currently not supported.")
```

File: art/attacks/poisoning/feature_collision_attack.py [hunk 4]
```diff
@@ -295,9 +310,9 @@ def tensor_norm(tensor, norm_type: Union[int, float, str] = 2):  # pylint: disab
         return tf.norm(tensor, ord=norm_type)
 
     if tensor_type in torch_tensor_types:  # pragma: no cover
-        import torch
+        import torch.nn as nn
 
-        return torch.norm(tensor, p=norm_type)
+        return nn.MSELoss(reduction="sum")
```
Comment (beat-buesser) [inline art/attacks/poisoning/feature_collision_attack.py:315]: Replace with `torch.norm`.

File: art/estimators/classification/pytorch.py [hunk 5]
```diff
@@ -811,13 +811,99 @@ def loss_gradient(  # pylint: disable=W0221
...
+        return grads
+
     def get_activations(
         self,
         x: Union[np.ndarray, "torch.Tensor"],
         layer: Optional[Union[int, str]] = None,
         batch_size: int = 128,
         framework: bool = False,
     ) -> np.ndarray:
+
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:906]:
```suggestion
```

File: art/estimators/classification/pytorch.py [hunk 6]
```diff
@@ -847,11 +933,12 @@ def get_activations(
 
         else:  # pragma: no cover
             raise TypeError("Layer must be of type str or int")
-
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:850]: Please don't remove this line.

File: art/estimators/classification/pytorch.py [hunk 7]
```diff
@@ -847,11 +933,12 @@ def get_activations(
...
         if framework:
             if isinstance(x, torch.Tensor):
                 return self._model(x)[layer_index]
-            return self._model(torch.from_numpy(x).to(self._device))[layer_index]
+            x_ = torch.from_numpy(x_preprocessed)
+            self._model(x_.to(self._device))
+            return x_, self._model._features[layer]
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:941]: Are these changes required?

File: art/estimators/classification/pytorch.py [hunk 8]
```diff
@@ -1037,14 +1137,9 @@ def get_layers(self) -> List[str]:
                         import torch.nn as nn
 
                         result = []
-                        if isinstance(self._model, nn.Sequential):
```
Comment (beat-buesser) [inline art/estimators/classification/pytorch.py:1040]: Does this new version also support `nn.Sequential` models?

File: art/attacks/poisoning/feature_collision_attack.py (hunk 4 above)
Comment (monshri) [reply art/attacks/poisoning/feature_collision_attack.py]: Done

File: art/estimators/classification/pytorch.py (hunk 7 above)
Comment (monshri) [reply art/estimators/classification/pytorch.py:940]: since, we are using hooks, we first need to pass the input through the model, which creates evaluation hooks at every layer and then we index that layer to get output. We can probably remove x_ (that is not required)

File: art/estimators/classification/pytorch.py (hunk 6 above)
Comment (monshri) [reply art/estimators/classification/pytorch.py:850]: Done

File: art/estimators/classification/pytorch.py (hunk 8 above)
Comment (monshri) [reply art/estimators/classification/pytorch.py:1040]: The granularity is maintained at module level, so i think it should.

File: art/estimators/classification/pytorch.py [hunk 9]
```diff
@@ -848,10 +925,29 @@ def get_activations(
...
+            interim_layer = dict([*self._model._model.named_modules()])[  # pylint: disable=W0212,W0622,W0613
+                self._layer_names[layer_index]
+            ]
+            interim_layer.register_forward_hook(get_feature(self._layer_names[layer_index]))
+
         if framework:
             if isinstance(x, torch.Tensor):
-                return self._model(x)[layer_index]
-            return self._model(torch.from_numpy(x).to(self._device))[layer_index]
+                self._model(x)
```
Comment (keykholt) [inline art/estimators/classification/pytorch.py:946]: Shouldn't this also use x_preprocessed instead of x?

Comment (monshri) [reply art/estimators/classification/pytorch.py:946]: I tried to maintain the same consistency as base code, which handles framework=True, this way. Please check - https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/dev_1.9.0/art/estimators/classification/pytorch.py

Comment (keykholt) [reply art/estimators/classification/pytorch.py:946]: Sure thing. Seems like it was addressed by PR 1471

**Comment by beat-buesser** [comment]
@monshri Could you please remove the two .pt files from this pull request?

**Comment by beat-buesser** [comment]
@monshri Could you please update the `notebooks/README.md` with the new notebook links and descriptions?

**Review by beat-buesser — APPROVED** [review]
Hi @monshri Thank you very much for implementing the feature collision poisoning attack in PyTorch! It looks good to me.
